### ISSUE TITLE
Erroneous values for derivatives of x*(1-x^a)

### ISSUE THREAD

**Issue by ianhbell** [autodiff/autodiff#242]
I define the function f(x) = x*(1-x^(1.1)). It is twice differentiable at x=0.  Above that, not.  At live.sympy.org I did:
```
>>> gamma = 1.1
>>> for n in range(10): print(n, diff(x*(1-pow(x, gamma)),x,n).subs(x,0))
(0, 0)
(1, 1)
(2, 0)
(3, zoo)
(4, zoo)
(5, zoo)
(6, nan)
(7, zoo)
(8, zoo)
(9, zoo)
```
When I calculate derivatives with Reals, I get the right first two derivatives, but zero for the higher derivatives, and I should get something nan-like. With Dual, all the returned values are nan, but I don't get the right first two derivatives. I *do* get the right derivatives if I only do two derivatives

Here is a snippet of the problem:
```c++
TEST_CASE("pure water derivatives ", "[NH3H2O]") {
    auto f = [](const auto& x) { return forceeval(x*(1 - pow(x, 1.1))); }; 
    
    autodiff::Real<6, double> x_ = 0.0;
    auto ders = derivatives(f, along(1), at(x_));
    CHECK(std::isfinite(ders[0]));

    using adtype = autodiff::HigherOrderDual<5, double>;
    adtype x__ = 0.0;
    auto dersdu = derivatives(f, wrt(x__,x__,x__,x__), at(x__));
    CHECK(std::isfinite(dersdu[0]));
}
```

**Comment by longemen3000** [comment]
how does it compare to `x - pow(x,1+a)` ?

**Comment by ianhbell** [comment]
Good idea, tried it, same result.

**Comment by allanleal** [comment]
Hi @ianhbell , thanks for reporting this issue. I'll take a look. The algorithms behind dual and real are different, but I think the culprit lies in the `pow` related codes. We'll see.

**Comment by allanleal** [comment]
Okay, I quickly checked here and it is indeed related to `pow(x)`. Here's what's happening. The algorithm behind `Dual` is recursive*, and for performance reasons, when evaluating `pow(x)` and its higher-order derivatives:

* only a single call to `std::pow` is performed (the intent here is to avoid multiple calls to a very expensive operation);
* this evaluation actually happens on the higher order derivative (i.e. `pow(x, 1.1)` is actually obtained indirectly from the evaluated derivatives, like doing `x*pow(x, 0.1)` in the case of the 1st order derivative)
* for a second-order derived evaluation, you end up somewhere in an evaluation of `std::pow(x, 1.1 - 1 - 1)`, which produces `inf`
* later operations involving `inf` (e.g. `0 * inf`) produce `nan`
* these `nan` values ​​are propagated upper to the chain calls and fully fill the resulting array with derivatives.

It is possible to include checks throughout the evaluations and try to remedy such behavior for such singular cases. However, I am more inclined not to do so because the main purpose of `autodiff` is to be fast for scientific applications, and such checks would certainly compromise performance. 

At least in my experience, I always know these singular cases in advance, and I try to stay away from them (e.g. variables representing amounts of chemical species must not be zero because I know at some point an evaluation of `log(x= 0)` will happen when calculating their chemical equilibrium state). In your case, you would prevent `x` from being zero by imposing a lower bound, say, `1e-16`. `autodiff` will be accurate to machine precision for evaluations at non singular points.

I feel that introducing such checks everywhere in the library (not just for `pow`, but also `log`, `exp`, `sqrt`, etc.) just to be accurate for singular cases is not a strong motivator. 

Thoughts?

****Note**: recursive expressions are generated at compile time also for performance reasons.*

**Comment by ianhbell** [comment]
I can certainly see why this happens based on your explanation and is what I had guessed before I submitted the issue.  I wonder though about what to do about it. I am concerned that I was getting the wrong answer without any indication that something went wrong. Particularly this is problematic in the ``Real`` case where I get 0 in the place of nan. For ``Dual``, the fact that all the results are ``nan`` provides an indication that "something bad happened" that can be traced back to the singular evaluation.  I think my preference would be also to pollute the ``Real`` results with ``nan`` for all the derivatives so that there is an indication that you need to take care.

One does not always know where the bad points are, or it can be very, very subtle. In my case I wasn't thinking about there being a potential problem. In this particular case I can just do as you propose, and bump the zero to ``1e-15``.  But I think a more general treatment is probably needed as the evaluations get more "black-box".

**Comment by allanleal** [comment]
As of now, evaluating the function and its higher-order derivatives at such singular cases produces *undefined results* (e.g., it could be `inf`, `nan` or even `zero` in case you get `some_num/inf`). 

Ideally, lower order derivatives that exist at some point should not be compromised because of higher order derivatives that do not exist at that same point. In case of `pow`, this would require reversing the recursive order of the expressions computing the derivatives. However, this can have a performance impact. For example, this change will then require division operations that do not exist in the current implementation (involving only multiplication, which is less costly). 

I'll keep this issue open for now, as I'd expect that improvements can be made in the near future for improved behavior at singular points. For now, avoiding evaluations at such points would be recommended by changing, e.g., zero by 1e-16 or some other small value.

**Comment by ianhbell** [comment]
Undefined is perhaps the worst possibility, could you at least make the undefined be ``nan`` until we come up with a better solution?

**Comment by allanleal** [comment]
This will require additional checks and inspections, which will compromise performance, and thus not something I'd like to add now. Revising the algorithm without compromising performance is the way to go, but I cannot work on this now. Just have some patience. :) 

BTW, it is not that difficult to know beforehand points to avoid (e.g., if your code involves `log`, stay away from zero and negative numbers, the same for `pow`, `sqrt`, etc.).

**Comment by ianhbell** [comment]
I agree about ``log``. It is evident because the function itself is not defined at ``x=0``.

The problem is that ``pow(x, 2.3)`` has two valid derivatives at 0, so it is not totally obvious that something bad is about to happen when you take the third derivative.  In fact this is precisely the problem I was having. The first derivative of ``x*pow(x,1.1)`` was fine, but not the second.

**Comment by allanleal** [comment]
I understand your point and agree that some cases can be less evident (e.g., `pow(x, 9.3)` should have all the first 9 higher-order derivatives well defined at `x=0`, but right now if you try to compute 10 or more higher-order derivatives, the current algo will spoil these, because the highest-order derivative is used to compute the lower ones). A solution is to invert this order; performance should have some impact though (it remains to measure), given that divisions will be introduced. I'm then divided between 

(1) fastest as possible but not informative enough for singular cases or 
(2) still fast and more informative for such particular cases.

Thoughts?

**Comment by ianhbell** [comment]
Personally, speed wins, and I don't want to compromise on that front.

I would be fine with a solution that is just as fast, but undefined values are converted to something invalid. Or all values are.  The only real design constraint for me is that I want to have some indication that something bad happened.

**Comment by allanleal** [comment]
One option is to introduce `assert` calls, but you'll only be able to notice that "something bad happened" in a debug run.

**Comment by ianhbell** [comment]
I think that is a good starting point.  I'm doing debugging mostly in debug mode, or at least sometimes.
